segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/ExtS3IntegrationTest.java

```diff
@@ -0,0 +1,359 @@
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import lombok.SneakyThrows;
+import org.junit.After;
+import org.junit.Before;
+
+/**
+ * End-to-end tests for SegmentStore, with integrated Storage and DurableDataLog.
+ */
+public class ExtS3IntegrationTest extends StreamSegmentStoreTestBase {
+    //region Test Configuration and Setup
+
+    private static final int BOOKIE_COUNT = 3;
+    private BookKeeperRunner bookkeeper = null;
+    private String baseDir;
+
+    /**
+     * Starts BookKeeper.
+     */
+    @Before
+    public void setUp() throws Exception {
+        bookkeeper = new BookKeeperRunner(this.configBuilder, BOOKIE_COUNT);
+        bookkeeper.initialize();
+
+        String endpoint = "http://127.0.0.1:9020";
+        URI uri = URI.create(endpoint);
+        Properties properties = new Properties();
+        properties.setProperty("s3proxy.authorization", "none");
+        properties.setProperty("s3proxy.endpoint", endpoint);
+        properties.setProperty("jclouds.provider", "filesystem");
+        properties.setProperty("jclouds.filesystem.basedir", "/tmp/s3proxy");
+        baseDir = Files.createTempDirectory("exts3_wrapper").toString();
+
+        this.configBuilder.include(ExtS3StorageConfig.builder()
+                                                     .with(ExtS3StorageConfig.EXTS3_BUCKET, "kanpravegatest")
+                                                     .with(ExtS3StorageConfig.EXTS3_ACCESS_KEY_ID, "x")
+                                                     .with(ExtS3StorageConfig.EXTS3_SECRET_KEY, "x")
+                                                     .with(ExtS3StorageConfig.EXTS3_URI, "http://localhost:9020"));
+    }
+
+    /**
+     * Shuts down BookKeeper and cleans up file system directory
+     */
+    @After
+    public void tearDown() throws Exception {
+        bookkeeper.close();
+        FileHelpers.deleteFileOrDirectory(new File(baseDir));
+    }
+
+    //endregion
+
+    //region StreamSegmentStoreTestBase Implementation
+
+    @Override
+    protected ServiceBuilder createBuilder(ServiceBuilderConfig builderConfig) {
+        return ServiceBuilder
+                .newInMemoryBuilder(builderConfig)
+                .withCacheFactory(setup -> new RocksDBCacheFactory(builderConfig.getConfig(RocksDBConfig::builder)))
+                .withStorageFactory(setup -> {
+                    StorageFactory f = new ExtS3StorageFactory(
+                            setup.getConfig(ExtS3StorageConfig::builder), setup.getExecutor());
+                    return new ExtS3ListenableStorageFactory(f, builderConfig.getConfig(ExtS3StorageConfig::builder));
```

Comment:
why don't you use the same pattern as in `HDFSIntegrationTest` where a `ListenableStorageFactory` is created that wraps another storage factory?